### PR TITLE
[JENKINS-62197]   fix(credentials): fix cannot use item credentials

### PR DESCRIPTION
**Pull request by conan17**

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabHookCreator.java [hunk 1]
```diff
@@ -17,13 +17,15 @@
 import org.gitlab4j.api.models.ProjectHook;
 import org.gitlab4j.api.models.SystemHook;
 
+import static io.jenkins.plugins.gitlabbranchsource.helpers.GitLabHelper.apiBuilder;
```
Comment (LinuxSuRen) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabHookCreator.java:20]: I was wondering if you have a strong reason why using a static import statement at here.

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabSCMSource.java [hunk 2]
```diff
@@ -720,10 +719,13 @@ public void afterSave() {
         GitLabHookCreator.register(this, webhookMode, systemhookMode);
     }
 
-    public PersonalAccessToken credentials() {
-        return CredentialsMatchers.firstOrNull(lookupCredentials(PersonalAccessToken.class, getOwner(),
-                Jenkins.getAuthentication(), fromUri(getServerUrlFromName(serverName)).build()),
-                GitLabServer.CREDENTIALS_MATCHER);
+    public StandardCredentials credentials() {
```
Comment (LinuxSuRen) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabSCMSource.java:722]: Changing a signature of a method is not encourage. Especially this is a public method, other users might already are using it in their own plugin.

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabHelper.java [hunk 3]
```diff
@@ -12,18 +20,45 @@
...
-            if (credentials != null) {
-                return new GitLabApi(server.getServerUrl(), credentials.getToken().getPlainText());
+            credentials = credentials == null ?server.getCredentials(): credentials;
+
+            if (credentials == null) {
+                return new GitLabApi(server.getServerUrl(), GitLabServer.EMPTY_TOKEN);
+            } else if (credentials instanceof PersonalAccessToken) {
+                return new GitLabApi(server.getServerUrl(),((PersonalAccessToken) credentials).getToken().getPlainText());
+            } else if (credentials instanceof UsernamePasswordCredentials) {
+                return new GitLabApi(server.getServerUrl(),((UsernamePasswordCredentials) credentials).getPassword().getPlainText());
```
Comment (LinuxSuRen) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabHelper.java:36]: So, this API only takes password? IMO, we can use `PersonalAccessToken` directly.

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabSCMSource.java (hunk 2 above)
Comment (baymac) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabSCMSource.java:722]: @conan17 You can add deprecate the old method and add your new one. For ref see how it is done in https://github.com/jenkinsci/github-branch-source-plugin/blob/master/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java#L667

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabHelper.java (hunk 3 above)
Comment (baymac) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabHelper.java:36]: I don't think username password can be used to access GitLab API. If we are to support OAuth2Login we can call this method https://github.com/gitlab4j/gitlab4j-api/blob/master/src/main/java/org/gitlab4j/api/GitLabApi.java#L170

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabSCMSource.java (hunk 2 above)
Comment (conan17) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabSCMSource.java]: I kept this part of the code, which is consistent with the original

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabHelper.java (hunk 3 above)
Comment (conan17) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabHelper.java]: Yes, if GitLab API only use PersonalAccessToken, does the credential selection column in the Item only need to list PersonalAccessToken?

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabHookCreator.java (hunk 1 above)
Comment (conan17) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabHookCreator.java:20]: This static method is required. The creation of GitLab Api client needs to judge the credentials, there will be duplicate code

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabHelper.java (hunk 3 above)
Comment (baymac) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabHelper.java]: Yes, I think we should do that but there is a scope for adding oauth support if you are interested.

**Comment by baymac** [comment]
@conan17 I do not understand Item credentials completely. My understanding is it a user level credentials, please correct me if I am wrong. And how can I add an item credentials to test it.

Also if we were to allow item credentials why not just update the GitLabServer descriptor (i.e. doFillCredentialsId) to be able to list item credentials and use the server credentials in api builder like it was before? Wouldn't that require less change?

**Comment by baymac** [comment]
Oh! Wait I see I confused with the checkout credentials of GitLab Source that you have changed as credentials of GitLab Server. But API uses credentials of GitLab Server. So you want item credentials to access the API then the change should be made in GitLab Server. I am kind of confused. Would be better if you specify your use case of what you want item credentials for then I may suggest the way to do it.

**Pull request closed without merging**